### ISSUE TITLE
show recursive calls

### ISSUE THREAD

**Issue by johnyf** [johnyf/pyan#2]
If a function calls itself, then no self-loop appears in the generated graph.

**Comment by Technologicat** [comment]
I was thinking the same thing. My Python 3 fork now does this.

**Comment by johnyf** [comment]
Thanks for adding this capability! Indeed I see the self-loops when I use https://github.com/Technologicat/pyan/commit/70d7e08d2f1939944ee735aeb12a20e136e57504 to plot https://github.com/johnyf/dd/blob/8fa0334ce322935cf19f414f613e00e8f1b82a03/dd/bdd.py#L686.

I noticed a case where a self-loop is shown, but the call is actually to the parent class, in particular the statement `super(Operator, self).__init__(*arg, **kw)` is interpreted as a recursive call, to `Operator.__init__`, even though it isn't (this call is at https://github.com/johnyf/omega/blob/d63b1b3ad9d110748292400cae3d2d04788d8be8/omega/symbolic/fol.py#L489).

**Comment by Technologicat** [comment]
Ah, that's probably because the analyzer doesn't know about `super()`. It thinks `super()` is a regular function call to something it hasn't yet seen, and then sees `__init__()`, which in the enclosing scope is bound to this `__init__()`, so by lexical scoping...

The use of `super()` is rather common, so this needs to be fixed. Thanks for the report!

Since I'll probably forget how to fix this before I have the time to do it, documenting here:

First we need to construct a MWE and look at the output with `--very-verbose`, to see how the analyzer interprets the line calling `super()`.

If it first resolves `super()` (as I think it should do), this is easy-ish to fix. We just need to record the `bases` in `visit_ClassDef()` (maybe change the format of `self.class_stack` to include `bases` for each entry?), and essentially implement a very simple form of `super()` in `visit_Call()`, matching by the function name. It can then analyze the arguments to find which base is named (or pick a sane default if not named), and then look up the `Node` corresponding to that class. The attribute lookup, already implemented, then does the rest.

**Comment by Technologicat** [comment]
Very rudimentary support for `super()` added in https://github.com/Technologicat/pyan/commit/6d61ad0ee334d48bfeb8f85b7891df73e3e7f665. Try it now?

Also, now supports inherited attributes (by looking them up in ancestors, not by wildcards as the old Pyan did).

In both cases, MRO is still very probably wrong, might look at that much later :)

**Comment by johnyf** [comment]
I tried it, the self-loop at the `__init__` node is still present.

**Comment by Technologicat** [comment]
Ok, thanks for testing.

Just now, I pushed https://github.com/Technologicat/pyan/commit/b8ec593508c5497133ee89b5ce0ada6c9ec98e90, which implements MRO with C3 linearization, based on the static type. If that doesn't help, I'll have to look at this more closely later.

**Comment by johnyf** [comment]
I tried with https://github.com/Technologicat/pyan/commit/cde6b3e25ca95e53b1a9af248b808441ea8a435b, which raised the following exception:

```python
./pyan.py --annotated /path/to/omega/omega/symbolic/fol.py --dot | dot -T pdf > foo.pdf
Traceback (most recent call last):
  File "./pyan.py", line 11, in <module>
    sys.exit(main())
  File "/.../pyan/pyan/main.py", line 105, in main
    v = CallGraphVisitor(filenames, logger)
  File "/home/ioannis/github/pyan/pyan/analyzer.py", line 148, in __init__
    self.process()
  File "/.../pyan/pyan/analyzer.py", line 155, in process
    self.process_one(filename)
  File "/.../pyan/pyan/analyzer.py", line 169, in process_one
    self.visit(ast.parse(content, filename))
  File "/.../python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/.../pyan/pyan/analyzer.py", line 347, in visit_Module
    self.generic_visit(node)  # visit the **children** of node
  File "/.../python3.5/ast.py", line 253, in generic_visit
    self.visit(item)
  File "/.../python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/.../pyan/pyan/analyzer.py", line 390, in visit_ClassDef
    self.visit(stmt)
  File "/.../python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/.../pyan/pyan/analyzer.py", line 454, in visit_FunctionDef
    self.visit(stmt)
  File "/.../python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/.../python3.5/ast.py", line 255, in generic_visit
    self.visit(value)
  File "/.../python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/.../pyan/pyan/analyzer.py", line 1020, in visit_Call
    self.visit(node.func)
  File "/.../python3.5/ast.py", line 245, in visit
    return visitor(node)
  File "/.../pyan/pyan/analyzer.py", line 742, in visit_Attribute
    obj_node,attr_node = self.get_attribute(node)
  File "/.../pyan/pyan/analyzer.py", line 665, in get_attribute
    obj_node,attr_name = self.resolve_attribute(ast_node)
  File "/.../pyan/pyan/analyzer.py", line 633, in resolve_attribute
    obj_node = self.resolve_builtins(ast_node.value)
  File "/.../pyan/pyan/analyzer.py", line 999, in resolve_builtins
    result = self.mro[class_node][1]
IndexError: list index out of range
```

Btw, `#!/usr/bin/python3` could be replaced with `#!/usr/bin/env python` at https://github.com/Technologicat/pyan/blob/cde6b3e25ca95e53b1a9af248b808441ea8a435b/pyan.py#L1.

**Comment by Technologicat** [comment]
Aaaaah, I see I didn't account for the corner case where the MRO has no bases. I'll need to fix this ASAP, as this may easily crash depending on the data thrown at it. Thanks!

Curious, though, why that would happen - isn't super() used only in cases where there is at least one other entry in the MRO beside the class itself? Maybe there's a base Pyan isn't seeing, so it gets left out of the MRO?

(The way I'm currently building the MRO is to first capture the AST nodes for the bases in pass 1. Between the passes, we then map all the captured AST nodes to Pyan's Node objects by lookup in the relevant lexical scope, and finally C3 linearize that for each class Pyan knows of. This populates `self.mro` for pass 2.)

Or maybe something is going wrong with the analysis. I'll have to look at this more closely.

I think I'll next download your repo and trace what is happening, as soon as I have time (some teaching to do, and the project for which I needed Pyan3 in the first place...).

The hashbang was generated by my installation of Spyder... still running on an old Linux that uses Python 2 by default :)

But yes, we could change it.

**Comment by Technologicat** [comment]
https://github.com/Technologicat/pyan/commit/eb041a9a45e2b54cbf397d5be8f2f5f27f089281 fixes this crash, and https://github.com/Technologicat/pyan/commit/05eafab802f4a0a6bcd6c86181d98539e991c761 fixes another one. I also changed the hashbang to use `/usr/bin/env` instead of hardcoding the path of `python3` - nice catch, in my previous comment I only noticed the `python` vs. `python3` difference.

Finally, there are some enhancements, see https://github.com/Technologicat/pyan/commits/master for details.

Using https://github.com/Technologicat/pyan/commit/89f71815087894c8e78978324b107d7a39db3b10 , which is the latest as of this writing, the problem analyzing `Operator.__init__` in `fol.py` is the following:

```
Call <_ast.Call object at 0x7f4cdc74f0f0>.__init__
Attribute __init__ of <_ast.Call object at 0x7f4cdc74f0f0> in context <class '_ast.Load'>
Resolve <_ast.Call object at 0x7f4cdc74f0f0>.__init__ in context <class '_ast.Load'>
Resolving super() of <Node class:fol.Operator>
Unresolved function call as obj, returning attr __init__ of unknown
Use from <Node method:fol.Operator.__init__> to <Node ???:*.__init__> (target obj <_ast.Call object at 0x7f4cdc74f0f0> not resolved; maybe fwd ref, function argument, or unanalyzed import)
New edge added for Use from <Node method:fol.Operator.__init__> to <Node ???:*.__init__> (target obj <_ast.Call object at 0x7f4cdc74f0f0> not resolved; maybe fwd ref, function argument, or unanalyzed import)
```
(Options: `./pyan.py fol.py --uses --no-defines --colored --annotated --dot -V >temp.dot 2>temp.log`)

Given just `fol.py`, Pyan does not see `_bdd.Function` (which, by the import on line 30, should be `cudd` in `dd.py`), so the `super()` call is not resolved in the analysis. This leads the analyzer to add a wildcard for any `__init__`, which then (incorrectly) resolves to `Operator.__init__`.

Maybe the solution is to simply ignore, and not add a wildcard when a `super()` call is not resolved...

**Comment by Technologicat** [comment]
The particular case of a spurious recursive uses edge to `Operator.__init__` when analyzing `fol.py` is fixed by https://github.com/Technologicat/pyan/commit/4f43527bf61260ee71e95a6af7a75cad1781a082 .

The solution was to detect failed `super()` lookups, not adding wildcards in that case (as the most likely reasons for such failures are that it is still pass 1, or some relevant source file is not in the analyzed set).

**Comment by johnyf** [comment]
Thanks for updating the hashbang. Indeed, the `bin/env` was the most important change (since `python3` is present by default). I think that I noticed the shebang when running in my Python 3 `virtualenv` raised an error, because it was calling `/usr/bin/python`, which happened to be Python 2.

Thanks for addressing the issue. I confirm that with https://github.com/Technologicat/pyan/commit/3d5cfe841538c6de30210bb687881d71a31a3eef no self-loop appears at the node of the method `Operator.__init__`.

Perhaps for failed lookups a leaf node could be added with the name of the `super` that wasn't found by the lookup, and maybe a question mark or some other indication that this is the boundary of the search. On the one hand that could induce clutter (I'm imagining omitting "failed" lookups from the standard library), but it may also provide an indication that there were failed lookups, so that additional paths should be included to get more complete results. I understand that the intended boundary is difficult to identify.

Also, to provide some context, as this use case brings up alternative imports. Strictly speaking, a static analyzer cannot know which candidate in:

https://github.com/johnyf/omega/blob/d63b1b3ad9d110748292400cae3d2d04788d8be8/omega/symbolic/fol.py#L29-L32

will fail, and which will succeed. Assuming that `pyan` finds both present in its search path, it may make sense to draw *two* outgoing edges for a single call. The subtrees may differ. Returning to this particular case, `dd` is a package which includes alternative implementations that expose the same API. The pure Python module `dd.autoref` is present in all installations, whereas the Cython module `dd.cudd` offers an interface to the C library [CUDD](http://vlsi.colorado.edu/~fabio/), and is present only if the library and the module are compiled (as of `dd == 0.5.3`, `manylinux` wheels are [available on PyPI](https://pypi.org/project/dd/0.5.3/#files) that contain `dd.cudd` compiled, with CUDD statically linked). The class `_bdd.Function` is either of:

1. `cudd.Function`: https://github.com/johnyf/dd/blob/8fa0334ce322935cf19f414f613e00e8f1b82a03/dd/cudd.pyx#L1326
2. `autoref.Function`: https://github.com/johnyf/dd/blob/8fa0334ce322935cf19f414f613e00e8f1b82a03/dd/autoref.py#L231

Btw https://github.com/davidfraser/pyan/commit/a2ff0655362fdedcf91e89f1a80b7e4b2464c444 and https://github.com/Technologicat/pyan/commit/3d5cfe841538c6de30210bb687881d71a31a3eef have diverged after the recent merge, rebasing may be needed.

**Comment by Technologicat** [comment]
Thanks for confirming! I tagged `v1.0.1` and sent a PR to David. The divergence was only due to David adding a note about where to find the old Python 2 Pyan into the README; I merged that change into my fork, and now it's mergeable again.

I also thought about adding nodes for failed lookups, and reached essentially the same conclusions you did: clutter caused by things that were not fully resolved, the standard library problem, and that identifying the boundary is difficult. The current solution is at least simple to explain and behaves predictably. :)

(About the standard library, Pyan has no idea of what is in it, and I'd like to keep it that way, to keep the analyzer both as simple as possible and version-agnostic. The analyzer is already becoming bloated, but I suppose that's just a consequence of the need to analyze programs with mutable state and OOP.)

Good point about alternative imports. Currently, by design, Pyan reads the source line by line, and uses the most recent binding in the current scope. The other option, to consider a probability-1/N use to each of the possible bindings, requires support for list-valued nodes (and in a different sense than actually assigning a list to a name). I might or might not add that later; we'll see.

Support for analyzing Cython source files is also something that is currently missing. I have some Cython projects of my own such as [pydgq](https://github.com/Technologicat/pydgq) and [wlsqm](https://github.com/Technologicat/python-wlsqm); so far, I still have a faint recollection of their internals, but the moment I don't, I'll need to add Cython support to Pyan :)

Maybe we could close this issue, and open new ones to track the particular new feature suggestions?
